server/src/main/java/org/elasticsearch/cluster/coordination/StableMasterHealthIndicatorService.java

```diff
@@ -0,0 +1,208 @@
+    }
+
+    @Override
+    public HealthIndicatorResult calculate(boolean includeDetails) {
+        HealthStatus stableMasterStatus;
+        String summary;
+        Map<String, Object> details = new HashMap<>();
+        Collection<HealthIndicatorImpact> impacts = new ArrayList<>();
+        MasterHistory localMasterHistory = masterHistoryService.getLocalMasterHistory();
+        if (hasSeenMasterInLast30Seconds()) {
+            long masterChanges = getNumberOfMasterChanges(localMasterHistory);
+            logger.trace("Have seen a master in the last 30 seconds");
+            if (localMasterHistory.hasSameMasterGoneNullNTimes(4)) {
+                DiscoveryNode master = localMasterHistory.getMostRecentNonNullMaster();
+                logger.trace("One master has gone null 4 or more times recently: " + master);
+                boolean localNodeIsMaster = clusterService.localNode().equals(master);
+                List<DiscoveryNode> remoteHistory = localNodeIsMaster
+                    ? localMasterHistory.getImmutableView()
+                    : masterHistoryService.getRemoteMasterHistory(master);
+                if (remoteHistory == null || MasterHistory.hasSameMasterGoneNullNTimes(remoteHistory, 4)) {
+                    if (remoteHistory == null) {
+                        logger.trace(String.format(Locale.ROOT, "Unable to get master history from %s", master));
+                    } else {
+                        logger.trace(String.format(Locale.ROOT, "The master node %s thinks it is unstable", master));
+                    }
+                    stableMasterStatus = HealthStatus.YELLOW;
+                    summary = String.format(
+                        Locale.ROOT,
+                        "The cluster's master has alternated between %s and no master multiple times in the last 30 minutes",
+                        master
+                    );
+                    impacts.add(
+                        new HealthIndicatorImpact(
+                            3,
+                            "The cluster is at risk of not being able to create, delete, or rebalance indices",
+                            List.of(ImpactArea.INGEST)
+                        )
+                    );
+                    if (includeDetails) {
+                        details.put("current_master", new DiscoveryNodeXContentObject(localMasterHistory.getCurrentMaster()));
+                    }
+                } else {
+                    logger.trace(
+                        String.format(
+                            Locale.ROOT,
+                            "This node thinks the master is unstable, but the master node %s thinks it is stable",
+                            master
+                        )
+                    );
+                    stableMasterStatus = HealthStatus.GREEN;
+                    summary = "The cluster has a stable master node";
+                }
+            } else if (masterChanges > 3) {
```

Comment:
To better match what the code does with our diagram plan - should this be the first thing we check?

Reply:
I put it here because technically a master going null and back four times is more than 3 master changes, so putting it here avoids running this logic for a master flapping null. But I can find a way to do this first if we need to stick to the diagram order.